[Feature] Add support for accented characters during passphrase input

abhigit-saha opened the pull request:
## Description

This PR introduces support for accented characters (like é, ä etc) during entry of the passphrase after completing the mnemonic entry flow. The seed bytes/fingerprints are verified by both Ian Coleman's BIP39 tool and python-mnemonic library by Trezor. 
<img width="1785" height="1722" alt="image" src="https://github.com/user-attachments/assets/829daa62-39f0-44d8-ac3a-9580cf20a189" />
![IMG20250711012107 (1) (1) (1)](https://github.com/user-attachments/assets/74898a51-07d3-4951-baa5-83edc6c9fb7e)



This pull request is categorized as a:

- [x] New feature
- [ ] Bug fix
- [ ] Code refactor
- [ ] Documentation
- [ ] Other

## Checklist

- [x] I’ve run `pytest` and made sure all unit tests pass before sumbitting the PR

If you modified or added functionality/workflow, did you add new unit tests?

- [ ] No, I’m a fool
- [x] Yes
- [ ] N/A

I have tested this PR on the following platforms/os:

- [x] Raspberry Pi OS [Manual Build](https://github.com/SeedSigner/seedsigner/blob/dev/docs/manual_installation.md)
- [ ] [SeedSigner OS](https://github.com/SeedSigner/seedsigner-os) on a Pi0/Pi0W board
- [ ] Other

abhigit-saha committed:
Added accented character support to passphrase screen

abhigit-saha committed:
Updated tests

notTanveer commented:
the final display on the input screen looks kinda jagged. for example, this "Û" — there's no clear separation between the "U" and the "^". Same thing with "S" and a few other uppercase letters. The lowercase ones look fine, but the uppercase ones are a bit rough around the edges.

abhigit-saha commented:
> the final display on the input screen looks kinda jagged. for example, this "Û" — there's no clear separation between the "U" and the "^". Same thing with "S" and a few other uppercase letters. The lowercase ones look fine, but the uppercase ones are a bit rough around the edges.

Noted! Also what are your views on language specific accented characters? Instead of having only a few accented characters (which are available everywhere), I think better would be to have small common set across all languages, and add  accented characters specific to the language in another keyboard

notTanveer commented:
honestly idk (but on a sidenote, sounds good). maybe post this up in the group.

alvroble commented on tests/test_seed.py:
```diff
@@ -7,13 +8,13 @@
 
 def test_seed():
-	seed = Seed(mnemonic="obscure bone gas open exotic abuse virus bunker shuffle nasty ship dash".split())
+	seed = Seed(mnemonic="obscure bone gas open exotic abuse virus bunker shuffle nasty ship dash".split(), passphrase="mühpassphrase")
```
Instead using a passphrase with only one accented character for test, I think is better to use one with all the supported characters ("áéíóúàèìòùâêîôûãõëïüÿçăâîșțÁÉÍÓÚÀÈÌÒÙÂÊÎÔÛÃÕËÏÜŸÇĂÂÎȘȚ") so we can be sure NFKD does not generate some byte inconsistency with some of those chars. Here I'd specifically look for a match with other software like Sparrow to ensure interoperability.

abhigit-saha replied:
> Instead using a passphrase with only one accented character for test, I think is better to use one with all the supported characters ("áéíóúàèìòùâêîôûãõëïüÿçăâîșțÁÉÍÓÚÀÈÌÒÙÂÊÎÔÛÃÕËÏÜŸÇĂÂÎȘȚ") so we can be sure NFKD does not generate some byte inconsistency with some of those chars. Here I'd specifically look for a match with other software like Sparrow to ensure interoperability.

Sparrow wallet does not put a restraint on what accented chars we are  using because they provide an empty field for the input of the same. We ofc cannot afford to do that since we have limited space to work with

alvroble commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -692,6 +695,8 @@ def __post_init__(self):
         # Isolate the more math-oriented or just uncommon symbols
         keys_symbol_2 = """^*[]{}_\\|<>/`~"""
 
+        keys_accented_lower = "áéíóúàèìòùâêîôûãõëïüÿçăâîșț"
```
These are missing `ñøåłđğşžčšœæþð`

alvroble commented:
> Noted! Also what are your views on language specific accented characters? Instead of having only a few accented characters (which are available everywhere), I think better would be to have small common set across all languages, and add accented characters specific to the language in another keyboard

Maybe if we focus just on latin-based alphabets for now, a practical approach could be to have one or two dedicated keyboards for accented characters. For example, one for accented vowels and another for consonants. I think making the keyboards fully language-specific would likely require a significant refactor to support non-latin languages.

abhigit-saha commented:
> > Noted! Also what are your views on language specific accented characters? Instead of having only a few accented characters (which are available everywhere), I think better would be to have small common set across all languages, and add accented characters specific to the language in another keyboard
> 
> Maybe if we focus just on latin-based alphabets for now, a practical approach could be to have one or two dedicated keyboards for accented characters. For example, one for accented vowels and another for consonants. I think making the keyboards fully language-specific would likely require a significant refactor to support non-latin languages.

Hi, sorry for the late reply. Currently there are two keyboards for the passphrase characters. One has lowercase ones and the other has uppercase ones. Now, both of these keyboards are already mostly at their maximum limit of characters (ie 27 characters) before the screen cannot handle anymore. Thus I had to leave out some other commonly used characters.

abhigit-saha commented:
> the final display on the input screen looks kinda jagged. for example, this "Û" — there's no clear separation between the "U" and the "^". Same thing with "S" and a few other uppercase letters. The lowercase ones look fine, but the uppercase ones are a bit rough around the edges.

Hi, can you please share an ss. While testing on my device, I think the Uppercase accented characters look fine, and accents could be distinguished. If you are facing a problem specifically in this regards, have you tried installing libraqm? maybe that would help?

abhigit-saha committed:
Modified passphrase tests to be stricter

alvroble commented:
I think it makes more sense to split accented characters across several smaller keyboards rather than limiting a single one to just 27 characters and leaving characters out. This approach would allow us to cover a wider range of accented characters, organize them more logically (by vowels/consonants/ligatures), and make the system easier to scale as we support more languages.

alvroble commented on tests/test_flows_seed.py:
I’d suggest against replacing the existing regular character test cases with the accented ones, but instead adding additional tests for the accented passphrase, so we maintain coverage for both standard ASCII and special character scenarios. Replacing them would mean to lose coverage here.

abhigit-saha committed:
Added multiple keyboards for accented characters

abhigit-saha committed:
Separated tests for accented and non-accented passphrases

notTanveer commented on tests/test_flows_seed.py:
```diff
@@ -53,6 +53,27 @@ def test_passphrase_entry_flow(self):
             FlowStep(seed_views.SeedOptionsView),
         ])
+        
+    def test_accented_passphrase_entry_flow(self):
```
can we reduce the duplication here? it’s an exact copy of test_passphrase_entry_flow. maybe we can extract it into a helper and just pass the custom passphrase.

abhigit-saha replied:
hi @notTanveer, thank you for the review! I've fixed this.

newtonick commented:
cACK. I'm a bit worried this PR would allow a user to choose a passphrase using characters not available on other hardware signing devices, but beyond that, conceptually I support this addition. I still need to review and test.

abhigit-saha committed:
Reused existing test function for testing of accented passphrase

notTanveer commented on tests/test_flows_seed.py:
```diff
@@ -42,19 +47,18 @@ def test_passphrase_entry_flow(self):
             FlowStep(seed_views.SeedOptionsView),
         ])
 
-
```
revert. ensures two blank lines between top-level definitions.

notTanveer commented:
> Hi, can you please share an ss.


installing libraqm fixed the issue i mentioned earlier, but now it’s not rendering characters like “jpygq” correctly, as you can see in the screenshot below. essentially, the lower parts of the letters are getting cut off. it’s most noticeable with lowercase characters, but slightly visible with uppercase ones as well.

<details>
<summary>📸 view screenshots</summary>

![pass_entry](https://github.com/user-attachments/assets/0eec1f89-ebd4-4e69-bf94-4f66e92af1ce)  
![pass](https://github.com/user-attachments/assets/8dcf4cd2-8fe0-47b4-96a5-f1f2b1f4db90)

</details>

abhigit-saha committed:
Fixed bounding box to correct display accented chars in verify screen

abhigit-saha committed:
Fixed: Backspace issue with nfkd normalization and reordered keyboards

abhigit-saha committed:
Ensured two linespaces between each test

abhigit-saha commented:
Hi devs this PR is ready for review.

alvroble commented:
Thanks @abhigit-saha!! I did test your PR as of https://github.com/SeedSigner/seedsigner/pull/788/commits/c8954a3e1bde4c507b9fd84d310703a183a59061. Launched this error when entering passphrase, then at SeedReviewPassphraseView pressing EDIT, then deleting the passphrase and pressing the back button:

```shell
Traceback (most recent call last):
  File "/home/pi/seedsigner/src/seedsigner/controller.py", line 313, in start
    next_destination = next_destination.run()
  File "/home/pi/seedsigner/src/seedsigner/views/view.py", line 164, in run
    return self._run_view()
  File "/home/pi/seedsigner/src/seedsigner/views/view.py", line 159, in _run_view
    return self.view.run()
  File "/home/pi/seedsigner/src/seedsigner/views/seed_views.py", line 438, in run
    selected_menu_num = self.run_screen(
  File "/home/pi/seedsigner/src/seedsigner/views/view.py", line 112, in run_screen
    self.screen = Screen_cls(**kwargs)
  File "<string>", line 23, in __init__
  File "/home/pi/seedsigner/src/seedsigner/gui/screens/seed_screens.py", line 1320, in __post_init__
    for i in range(0, len(self.passphrase), chars_per_line):
ValueError: range() arg 3 must not be zero
```

I think you need to merge with dev, since there was a recent merge (#790). It might not cause conflicts at the merge level, but it could at the functional level. To be fair, this problem might have existed before that PR was merged… it’s a tricky one 😅. I tested after merging dev and the problem was indeed gone.

abhigit-saha committed:
Merge branch 'SeedSigner:dev' into passphrase-accented

abhigit-saha commented:
Its up to date now!

alvroble commented:
Tested ACK on Raspberry Pi OS Manual Build as of https://github.com/SeedSigner/seedsigner/pull/788/commits/ffaa392663c1d425f10c7464d364ea4a2a7e9f08.

Nice work here @abhigit-saha!!

notTanveer commented:
apparently, looks like the accent mark of C has moved to N. 

<details>
  <summary>images</summary>
  
![b](https://github.com/user-attachments/assets/f1fb9582-2fe5-455a-b6ba-e36a85e94886)
![a](https://github.com/user-attachments/assets/6f396270-6ccb-4abf-adcc-664e196989ea)

</details>

abhigit-saha committed:
Changed to passphrase_display to accurately split passphrase

abhigit-saha committed:
Merge branch 'passphrase-accented' of https://github.com/abhigit-saha/seedsigner into passphrase-accented

abhigit-saha commented:
> apparently, looks like the accent mark of C has moved to N.
> 
> images

Hi @notTanveer thanks a lot for testing with these edgecases so comprehensively! I have fixed this. The main issue was that in the Passphrase Review screen the passphrase was being passed to as the NFKD normalized version (ie accents are counted in the length), which would cause issues if we are splitting the passphrase into two lines based on length, since it could happen that the accent of a letter goes into a separate line. Changed it to NFC encoding while displaying (using the `passphrase_display` function that was previously unused throughout the codebase) :)

kdmukai commented:
I think this feature would require a Settings option to enable these extra keyboards.

Otherwise:
1. You're inconveniencing EVERY user that has no intention of ever including an accented char when they cycle through the keyboards. This would annoy the hell out of me.
2. If they're enabled by default, it encourages users to consider using accented chars in their passphrase. Which, as @newtonick mentions, is a serious concern about compatibility in the broader ecosystem.

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -692,6 +699,18 @@ def __post_init__(self):
+        
+        # Keyboard 3: Some common accented consonants
+        keys_accented_lower_3 = "çñşțćłšžčřňťđĺŕľĵĝħ"
+        keys_accented_upper_3 = "ÇÑŞȚĆŁŠŽČŘŇŤĐĹŔĽĴĜĦ"
```
I have no expertise here, but how do we know that these really are the "common" vs "rare" vowel and consonant sets? For all languages that use Latin chars? For the most common? Or is it an opinionated list biased from a particular point of view (e.g. coming from, say, French)? Is it even possible to make any kind of determination?

Also are the "æø" chars considered "accented" chars? They seem more like just additional chars that aren't in the standard Western Latin set. But really what this means is that this approach is trying to accommodate a language's full alphabet (e.g. Icelandic needs the accented versions plus the "æ").

Quick googling: Norweigian needs "æ", "ø", AND "å" -- ~~but that is NOT included. Swedish also needs "å".~~ (edit: "å" is in the list. My eye missed it)

German needs "äöüß".

These are very messy waters. Trying to be definitive about a character set is making me very uncomfortable right now.

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -861,6 +1006,30 @@ def _render(self):
+            self.hw_button1.text = self.KEYBOARD__ACCENTED_UPPERCASE_3_BUTTON_TEXT
+
+        elif self.initial_keyboard == self.KEYBOARD__ACCENTED_UPPERCASE_3_BUTTON_TEXT:
+            cur_keyboard = self.keyboard_accented_upper_3
```
With this much repetition, it feels like the initial settings should be set via indexing into a list or some kind of dict via a constant.

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -932,10 +1101,38 @@ def _run(self):
+                        cur_button1_text = self.KEYBOARD__ACCENTED_UPPERCASE_3_BUTTON_TEXT
+                    elif cur_button1_text == self.KEYBOARD__ACCENTED_UPPERCASE_3_BUTTON_TEXT:
+                        self.keyboard_accented_upper_3.set_selected_key_indices(x=cur_keyboard.selected_key["x"], y=cur_keyboard.selected_key["y"])
+                        cur_keyboard = self.keyboard_accented_upper_3
```
Similarly, this swap cycling feels like it should be defined in one simple dict/list/structure to avoid all this repetition.

The original implementation was only cycling 2-3 keyboards so the repetition wasn't quite worth refactoring away. But with this many swaps, it absolutely becomes worth it.

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -958,6 +1155,18 @@ def _run(self):
+                    elif cur_keyboard == self.keyboard_accented_lower_3:
+                        cur_button1_text = self.KEYBOARD__ACCENTED_LOWERCASE_3_BUTTON_TEXT
+                    elif cur_keyboard == self.keyboard_accented_upper_3:
+                        cur_button1_text = self.KEYBOARD__ACCENTED_UPPERCASE_3_BUTTON_TEXT
```
Yet another place where if we had one single list/dict/whatever, we could avoid all of this.

kdmukai commented on tests/test_flows_seed.py:
```diff
@@ -33,7 +33,13 @@ def test_scan_seedqr_flow(self):
-    def test_passphrase_entry_flow(self):
+    test_passphrases = [
+        "muhpassphrase",
+        "áéíóúàèìòùâêîôûãõëïüÿăąæøåðçñşțćłšžčřňťđĺŕľĵĝħÁÉÍÓÚÀÈÌÒÙÂÊÎÔÛÃÕËÏÜŸĂĄÆØÅÐÇÑŞȚĆŁŠŽČŘŇŤĐĹŔĽĴĜĦ"
```
It would be better to have these chars in vars or constants that can be referenced in the main code and here. That way we can be sure that no characters were missed in the test. Just too ugly a maintenance task to hard code the list in two different places.

kdmukai commented on tests/test_seed.py:
```diff
@@ -7,13 +8,23 @@
+	seed_1 = Seed(mnemonic="obscure bone gas open exotic abuse virus bunker shuffle nasty ship dash".split(), passphrase="muhpassphrase")
 	
-	assert seed.seed_bytes == b'q\xb3\xd1i\x0c\x9b\x9b\xdf\xa7\xd9\xd97H\xa8,\xa7\xd9>\xeck\xc2\xf5ND?, \x88-\x07\x9aa\xc5\xee\xb7\xbf\xc4x\xd6\x07 X\xb6}?M\xaa\x05\xa6\xa7(>\xbf\x03\xb0\x9d\xef\xed":\xdf\x88w7'
+	assert seed_1.seed_bytes == b'\x13\xc8\xdfn\x17\xf4\xb3\x06n\x12\xfd\x1bE\x0c\x8d\x819\xe2\xf1\x9ar\xc1j\xde\x14<\xe8\x98\x8e\xd8=\xb8\x94j\x9a\x0c\xaf\xf4\x04\xa6s\xc8\xb4\xd3\xf1\xd4\x85\xde\x84u\xac\x88\\p\xeb\xcf\xbd\xc4A\xb5\xa1\xc3\xb1\xfa'
```
It's a shortcoming of the original test code, but we should explicitly state how we generated this comparison / verification data (e.g. simply referencing iancoleman.io and mention if any special steps / settings are required in that tool).

kdmukai commented:
@abhigit-saha I encourage you to continue improving this PR per the comments above. The DRY refactoring in particular is an excellent exercise and maybe even a tiny bit challenging.

However, I am extremely skeptical about this as a feature. Even when this PR is in its ideal form, I'm not sure I will be voting in favor of merging it. My own lack of familiarity with these accented/extra characters is a huge blocker. Obviously that's my shortcoming and not your fault. It would help a lot to hear from bitcoin experts who operate in these languages. What do they consider the best practices around bip39 passphrases?

abhigit-saha commented:
> @abhigit-saha I encourage you to continue improving this PR per the comments above. The DRY refactoring in particular is an excellent exercise and maybe even a tiny bit challenging.
> 
> However, I am extremely skeptical about this as a feature. Even when this PR is in its ideal form, I'm not sure I will be voting in favor of merging it. My own lack of familiarity with these accented/extra characters is a huge blocker. Obviously that's my shortcoming and not your fault. It would help a lot to hear from bitcoin experts who operate in these languages. What do they consider the best practices around bip39 passphrases?

Thank you Keith for the reviews! Most self-custodial softwares I have interacted with have a textarea for the passphrase, meaning that if the user wants they can use whatever ascii recognizable characters they want. 

The situation for hardware wallets is indeed different. I've only ever interacted with Jade apart from Seedsigner, which doesn't support this. I would love other viewpoints as well. Although it's clear that it shouldn't be the default behavior and enabled using a setting

abhigit-saha committed:
Refactored multiple if stataements using dict

abhigit-saha committed:
Refactored furthur

abhigit-saha committed:
Added conftest.py file to have test constants

abhigit-saha committed:
Comments explaining how to generate seed added

abhigit-saha committed:
Changed existing implementation to only have special passphrase characters of the selected locale

abhigit-saha commented:
I have made the necessary changes suggested (here are the [commits](https://github.com/SeedSigner/seedsigner/compare/80534363a796e6ac6d6c12e06ede9e676f13cb9e~5...80534363a796e6ac6d6c12e06ede9e676f13cb9e)
I have also changed the keyboards to only include the charset of the current locale if they enable the setting to include the full charset, optionally. That way we avoid the ambiguity of including every single character available and instead add support for languages incrementally.  (I think inputs from the translators would be very valuable here!) 
The bip39 helper module is the same as created in this [PR #714 ](https://github.com/SeedSigner/seedsigner/pull/714), and since Passphrases were introduced in BIP39 it makes sense to club related helper functions in the same helper module. 

Additionally I would be changing the naming from `accented` to `special` since that is more appropriate.

abhigit-saha committed:
Changed references of full and accented to special

abhigit-saha committed:
Updated messages.pot

kdmukai commented on src/seedsigner/models/settings_definition.py:
```diff
@@ -312,6 +312,17 @@ def map_network_to_embit(cls, network) -> str:
+        (LOCALE__SPANISH, ["áéíñóúü", "ÁÉÍÑÓÚÜ"]),
+        # Italian - Based on standard Italian orthography using acute and grave accents on vowels
+        (LOCALE__ITALIAN, ["àèéìíîòóùú", "ÀÈÉÌÍÎÒÓÙÚ"]),
+    ]
```
Ah, this is VERY interesting.

I have NOT thought deeply about this, but it's possible that someone might want to use the device with the UX in one language (e.g. English) but use a bip39 passphrase using words from a different language (e.g. someone knows just enough Spanish to use "Bitcoin-es-como-jalapeños").

Probably too much of an edge case to worry about. But if it is a serious concern, we'd probably need a separate setting for which charset to offer for bip39 passphrases. That's probably too much fussing. Though maybe such a setting could be like:
* Match selected language (default)
* Español
* Italiano
* etc...

kdmukai commented:
Also what about punctuation marks that we don't use in English?

```
¡¿
```

Not sure if there are others.

If the list is short (possibly literally just those two), then we should probably just add them to the existing symbols keyboard.
